src/Mvc/Mvc.Core/src/ApplicationParts/RelatedAssemblyAttribute.cs

```diff
@@ -106,7 +123,7 @@ public static IReadOnlyList<Assembly> GetRelatedAssemblies(Assembly assembly, bo
                     }
                 }
 
-                var relatedAssembly = loadFile(relatedAssemblyLocation);
+                relatedAssembly = loadFile(relatedAssemblyLocation);
```

Comment:
Also this LoadFile isn't correct as it loads the assembly into a different load context.  @vitek-karas What would we be using instead?  `AssemblyLoadContext.GetLoadContext(typeof(RelatedAssemblyAttribute)).LoadFormAssemblyFile`?

Reply:
It depends on what the goal is. If the usage of the related assemblies is to "attach" some code to an existing assembly, then I think it would make sense to load the related assemblies into the same load context as the "annotated" assembly.
So something like `AssemblyLoadContext.GetLoadContext(assembly).LoadFromAssemblyPath` (`assembly` is where we look for the attributes).

Note though that this may be a breaking change. `Assembly.LoadFile` has the specific behavior that it will load into a new load context for every file. Side effect of this is that it pretty much always succeeds (there will never be any versioning problems) - so if the related assemblies contain two assemblies of the same name but from different place and thus potentially different version, `LoadFile` will not fail on them. Loading into a manually picked ALC might fail (depends if there's any load context isolation in play or not).
That said `LoadFile` typically leads to more subtle versioning problems - even though the assembly specified is loaded in isolation and thus always works, its dependencies are NOT isolated and will fallback to the Default ALC. So for example if I have two assemblies which both use Newtonsoft.Json but one version 11 and the other version 12 and I load those two assemblies via `LoadFile`, it will likely lead to some kind of problem - either loading of references for one of them will fail (if 11 gets loaded before 12 is asked for), or I might get weird exceptions later on (if 12 is used for reference to 11 and there are behavioral changes in 12).

All in all - I think it would be great to also look in detail into what assembly loading behavior is desirable in this code. That said it's not really related or necessary for single-file use cases (those will pretty much work as before).

Reply:
It likely has no meaningful observable difference (other than avoiding a new context) as these assemblies have unique names and share references with the app

Reply:
To expand on @davidfowl's comment, these assemblies contain the result of compiling views that the application cannot reference. They're loaded so that MVC can discover and execute them. They also have the exact set of dependencies as the application so there's little risk of version conflict.

Reply:
In that case using `AssemblyLoadContext.GetLoadContext(assembly).LoadFromAssemblyPath(path)` is probably the best solution. It should work even if the code is using multiple ALCs (right now it probably never does), so good for the future and it saves you the additional ALCs. It also sort of guards against potential "misuse" which you don't want to support anyway. Having the related assemblies share the same ALC as the "main" assembly will also have the nice side-effect that they will share the dependency resolution with the "main" assembly. So if in the future you add some logic to how dependencies are resolved and so on, the related assemblies would just keep on working.